### ISSUE TITLE
The concordance Site Algorithm: Discussion

### ISSUE THREAD

**Issue by LinguList** [pylogeny/concordance#4]
So if I understand this correctly, the algorithm proceeds as follows:

1. use an unrooted tree
2. now iterate over all branches in this tree (note that the root is lost here!)
3. for each branch, split the tree in two parts along the branch, and make this a fourpartition by taking the first siblings of the following parts (this is different from the quartet partitions I had envisioned)
4. now pull out four random leave nodes here, along with their characters
5. determine the scenario: 
    - sites are not decisive and not concordant: i.e., `a == b == c == d` or `a != b != c != d`
    - sites are decisive for parsimony, but not concordant, meaning `a == b == c or a == c == d or b == c == d or (a == c and b == d)`, etc.
    - sites are concordant with the branch, meaning: `a == b and c == d and a != c`
6. divide for this trial the number of concordant sites b the number of decisive sites
7. take the mean of many of these samples

**Comment by LinguList** [comment]
This leads to a rather simple computation strategy, that can be based on the `as_graph` construct which I introduced in the `util.py` script.

1. turn tree to unrooted graph
2. iterate over all edges in the graph (easy graph manipulation task, we can even store the edges before)
3. use the `split_by_node` function to split the graph into two graphs, and then take the direct descendants of both nodes
4. use simple if-else constructs to determine the parsimony sites

I think, this can be done in a rather simple fashion, without any tree classes, apart from the one from which we derive the unrooted graph. If we want to then define this for rootet trees, we can do that later.

**Comment by WesScivetti** [comment]
> 
> 
> So if I understand this correctly, the algorithm proceeds as follows:
> 
>     1. use an unrooted tree
> 
>     2. now iterate over all branches in this tree (note that the root is lost here!)
> 
>     3. for each branch, split the tree in two parts along the branch, and make this a fourpartition by taking the first siblings of the following parts (this is different from the quartet partitions I had envisioned)
> 
>     4. now pull out four random leave nodes here, along with their characters
> 
>     5. determine the scenario:
>        
>        * sites are not decisive and not concordant: i.e., `a == b == c == d` or `a != b != c != d`
>        * sites are decisive for parsimony, but not concordant, meaning `a == b == c or a == c == d or b == c == d or (a == c and b == d)`, etc.
>        * sites are concordant with the branch, meaning: `a == b and c == d and a != c`
> 
>     6. divide for this trial the number of concordant sites b the number of decisive sites
> 
>     7. take the mean of many of these samples

This makes sense to me, this is how I understood the scf according to Minh et al. (2020)

**Comment by LinguList** [comment]
Cool, so I just pushed code to make the fourpartition using our new tree-class, the as_graph function (very simple) and a simple get_descendants function. this currently returns frozensets, but we can turn them to lists as well.

**Comment by LinguList** [comment]
Usage would be (installing the package):

```python
from pyconcordance.util import fourpartition, tree
tree = Tree("(((A,B)E1,(C,D)E3)E2,E)root;"
fourpartition(tree, "E1", "E2")
```

Output is: 

```
(frozenset({'A'}), frozenset({'B'}), frozenset({'E'}), frozenset({'C', 'D'}))
```

But we switch to lists then, for convenience.

this is the fourpartition for the edge that is between E2 and E1, in an unrooted version of the tree `tree`.

**Comment by LinguList** [comment]
```python
In [3]: print(tree.root.ascii_art())
                     ┌─A
              ┌─E1───┤
              │      └─B
       ┌─E2───┤
       │      │      ┌─C
──root─┤      └─E3───┤
       │             └─D
       └─E

```

**Comment by LinguList** [comment]
So note that there are not too many edges, where you can make fourpartitions of such a tree. But I am not smart enough now to tell, which edges these are (E1, E2 work, E2 E3, but none else). 

So the next step would be to compute all edges (which should be easy). And then to add a function that makes the selection. 

I'd start with a function to make the little parsimony based on if-else statements, this should be straightforward, a function call would be:

```python
def decisive(a, b, c, d):
    if a == b == c == d or a != b != c != d:
        return False
    return True

def concordant(a, b, c, d):
    if a == b and c == d and a != c:
        return True
    return False
```

**Comment by LinguList** [comment]
We should maybe make a small test dataset with lots of random characters for a simple tree of only 8 to 10 leaves, so we can check this. This can be put into a JSON/python dict structre in examples.py and loaded from there to develop this algorithm.

**Comment by PhyloStar** [comment]
> 5. determine the scenario:
>    
>    * sites are not decisive and not concordant: i.e., `a == b == c == d` or `a != b != c != d`

This would include a singleton character such as "AAAB" which is not decisive either. Definition of parsimony informative from Yang:

> For a site to be a parsimony-informative site, at least two characters have to be observed, each at least
twice.

`if "-" not in col and len(set(col)) == 2 and "?" not in col:` something like this?

**Comment by PhyloStar** [comment]
Sorry, I did not understand the concordant part in the original issue creation. We do the two possible  Nearest Neighbor interchange (NNI) to a quartet and check if the interchanged quartets have lower parsimony score than the original quartet's parsimony score.

Here col is the site (or column)
`pars_orig_score =  fitch_pars(col)`

NNI 1
`swap1 = [col[0], col[2], col[1], col[3]]
swap1_score = fitch_pars(swap1)`

NNI 2
`swap2 = [col[0], col[3], col[2], col[1]]
swap2_score = fitch_pars(swap2)`

**Comment by PhyloStar** [comment]
I get `ModuleNotFoundError: No module named 'pyconcordance'` for pyconcordance. I did run `python3 setup.py install` and `python3 setup.py develop`. Same error both times.

**Comment by LinguList** [comment]
It is called `pyloconcordance`.

**Comment by LinguList** [comment]
So our suffix will always be `pylo`.

**Comment by LinguList** [comment]
And thanks for the additional info on parsimony-informative sites, @PhyloStar. Means even simpler code.

**Comment by PhyloStar** [comment]
I think the definition in Minh et al. works since their data is DNA based. In our case, we get cloze to zero concordance scores because of the large number of symbols in the data. I think we need weighted parsimony here.

**Comment by PhyloStar** [comment]
Bad idea here. Parsimony informative is needed for weighted variant too. We will need a likelihood based concordance factor to prevent ignoring sites that are non-informative.

**Comment by PhyloStar** [comment]
> Bad idea here. Parsimony informative is needed for weighted variant too. We will need a likelihood based concordance factor to prevent ignoring sites that are non-informative.

Will open a different issue.

**Comment by LinguList** [comment]
Yes, I also encountered this, I started testing the concordance factor in the version I provided now in `concordance.py` on some actual cognate set data with NJ trees, and the factor goes sometimes down to 0, sometimes, I do not even encounter useful examples, but I had some cases where all were along 30% as they said would be an indicator of randomness. So all after all, this IS an interesting method alternative for bootstrapping, but it is not clear if we do it right, and what should be done.

**Comment by LinguList** [comment]
But coding this up with lingpy is very easy now. You need the tree, it can also be multifurcating, since I adapted your strategy, and you assemble the patterns via the lingpy wordlist:

```python
wl = Wordlist("Path2Wordlist")
tree = Tree("yourtree")
etd = wl.get_etymdict(ref="cogid")
patterns = []
for cogid, row in etd.items():
    pattern = []
    for idx in row:
        if idx != 0:
            pattern += ["1"]
        else:
            pattern += ["0"]
    patterns += [pattern]
scores = rscf(tree, taxa, patterns, iterate=100)
all_scores = []
for node in tree.preorder[1:]:
    if node.descendants:
        all_scores += [scores[node.name]["alld"]]

print('{0:20} | {1:.4f} | {2}'.format(
    cognate,
    sum(all_scores)/len(all_scores),
    len(patterns)
    ))
```

**Comment by LinguList** [comment]
And for multi-states (where it goes south):

```python
patterns = []
for concept in wl.rows:
    idxs = wl.get_dict(row=concept, entry="cogid")
    pattern = []
    for taxon in wl.cols:
        entry = idxs[taxon]
        if not entry:
            pattern += ["Ø"]
        else:
            pattern += [[str(e) for e in entry]]
    patterns += [pattern]
scores = rscf(stree, wl.cols, patterns, iterate=100)
all_scores = []
for node in stree.preorder[1:]:
    if node.descendants:
        if [n for n in node.descendants if n.descendants]:
            all_scores += [scores[node.name]["alld"]]
print('{0:20} | {1:.4f} | {2}'.format(
    cognate,
    sum(all_scores)/len(all_scores),
    len(patterns)
    ))
```

**Comment by LinguList** [comment]
Ah, forgot the import:

```python
from pyloconcordance.concordance import rscf
from pylotree import Tree
```

**Comment by LinguList** [comment]
@PhyloStar, this function is very rurdimentation, based on what we discussed, but taking slightly different paths for implementation from your version with dendropy. Could you test this and specifically see if I get the interpretation of concordance in this rooted version right? In my opinion, concordance in a rooted tree would say: you pick a node, to support this node, the sampled quartets (in binary trees) should have the structure AB vs. CD with the form of `aa bc`, while `aa ab` would not be good, and neither would `ab cd`.

As far as I see this now, for binary data, this may yield some results, but they are hard to interpret. We could sample partitions, like stable vs. unstable concepts. 

But before, it must be secured that the code is not doing rubbish.

**Comment by PhyloStar** [comment]
@LinguList I will test the code tomorrow and report the results. Thank you.

**Comment by PhyloStar** [comment]
One question about the parts C and D. Doesn't it have to computed explicitly?

**Comment by LinguList** [comment]
One could argue that it is enough to just sample two other nodes, so one could do it in an even simpler way.

**Comment by LinguList** [comment]
In fact, the root-based perspective changes really a lot. So we would define a branch-supporting concordant character as a character with states `A == B and A != C and A != D and B != C and B != D`. Since the descendants of the node can also be three at once, I used sets to determine character identity, but if it is about a single NODE, we could argue that we should be able to sample any two languages from the set of leave nodes, right?

**Comment by PhyloStar** [comment]
Yes. I think checking if A == B would be enough. It would be weak parsimony criteria. Even if C !=D, A == B would be enough to posit that A and B are grouped together. One question is what happens if C == D and A!=B. Need to check if this is working.

**Comment by LinguList** [comment]
This is the opposite case, where the characters are concordant, but 
decisive, right? I mean, they are decisive for a node with two 
descendants, but NOT concordant with the node of interest, which is AB.
